docs/v1/P4Runtime-Spec.mdk

```diff
@@ -1456,41 +1456,132 @@ types (32-bit and 64-bit words). The P4 language does not put any limit on the
 size of integer values, whether unsigned (`bit<W>`) or signed (`int<W>`), and it
 is up to the P4 programmer to choose the appropriate sizes. Because of this
 flexibility, P4Runtime represents P4 integer values as binary strings, using the
-bytes Protobuf type. The correct bitwidth - as per the P4 program - of each
-integer variable exposed through P4Runtime is specified in the P4Info message
-and it is up to the server to ensure that the binary string provided by the
-client has the correct length. The server must return an `INVALID_ARGUMENT`
-error code otherwise. For each integer value, the expected length in bytes is
-obtained by rounding-up the bitwidth to the nearest byte:
-`p4runtime_binary_string_length = (p4_bitwidth + 7) / 8`.
-
-For all binary strings, P4Runtime uses big-endian (&ie; network) byte-order. If
-the P4 integer type is not byte-aligned (`p4runtime_binary_string_length !=
-p4_bitwidth * 8`), P4Runtime expects the client to pad the value with unset bits
-before the most-significant bit of the value.
+`bytes` Protobuf type. The correct bitwidth - as per the P4 program - of each
+integer variable exposed through P4Runtime is specified in the P4Info message.
 
+The canonical binary string representation uses the shortest string that
+fits the encoded integer value.  This representation achieves three goals:
+
+* It ensures that a properly encoded binary string's integer value conforms
+  to the P4Info-specified bitwidth.
+
+* It supports [read-write symmetry](#sec-read-write-symmetry).
+
+* It helps facilitate non-disruptive P4 program updates.
+
+The shortest possible binary string for an integer value with `high_bit`
+as the most-significant non-zero bit in the value is computed as:
+~ Begin CPP
+encoded_string_size = (high_bit + 8) / 8;
```

Comment:
if the value is 0, what is `high_bit`?

Reply:
Good point - will address in next commit.